### ISSUE TITLE
Support kubenet for AKS Service with Windows nodes

### ISSUE THREAD

**Issue by previ** [Azure/AKS#1244]
Currently AKS service support Windows nodes with Azure CNI only.

Azure CNI mandates a flat network for pods, services and nodes, which is not always desirable/possible.
In my use case, due to some restrictions in vnet/subnets address space, an overlay network (such as the one provide by kubenet) is a mandatory requirement.

Request
Support kubenet (or other network CNI which provides overlay network with separate address space for pods and services) for AKS service with Windows nodes, too.

**Comment by MikkelHegn** [comment]
Is the main concern with Azure CNI, the reservation of ip-addresses on the cluster subnet? Or are there other requirements you have?

**Comment by previ** [comment]
> Is the main concern with Azure CNI, the reservation of ip-addresses on the cluster subnet? Or are there other requirements you have?

Yes the need to reserve a large, flat address space for pods/services is the main issue, since we have limited space to dedicate to AKS clusters and it is perfectly ok for our use case to have pods/services on an overlay, unreachable network.

**Comment by MikkelHegn** [comment]
We are working on reducing the reserved set of IPs for worker nodes (allowing to change the min. pod per node below 30). Would such a change suffice, or would you require overlay?

**Comment by previ** [comment]
The best solution would be the overlay since a flat address space for pods/services is mostly a problem for our use cases, rather than a feature.
Also, the flat address space prevents the reuse of the same addresses for different clusters.

**Comment by mvandewouw** [comment]
kubenet support for windows aks nodes would really help

**Comment by shahga** [comment]
This is really a dire need to embrace Windows containers. We have AKS Linux with Kubenet and we would like to scale that by adding Windows nodepools but we cannot, because Windows needs CNI. Secondly, the IP needs of CNI are just too greedy to be used in large enterprise.

**Comment by MikkelHegn** [comment]
@shahga - we only require 10 IPs per node to be reserved, is that too many for your scenario?

**Comment by shahga** [comment]
So 10 IPs per Node is certainly better than one IP for every Pod, but it can still be quite limiting. Our subnets need to bear burden of IPs for nodes, extra IPs nodes for upgrade scenarios, extra IPs for scaling, IPs for ILB for Kubernetes service fronting pods, IPs for private links and IPs other workloads, so a node taking up 10 IPs can be crushing.

**Comment by previ** [comment]
Same for our use case: we have a specific address space dedicated to azure and a few subnets dedicated to AKSes; even as low as 10 IPs per node can be high, if these IPs are reserved and not reusable by  other AKS instances.
As a matter of fact, we do not see the "one real IP per pod" as an advantage of any kind over an overlay network for pods, and expose only one single ingress controller via LoadBalancer service kind.

**Comment by aanandr** [comment]
@MikkelHegn - why do we need to reserve 10 IPs for every node?
@shahga , @previ - is allocation of a VNet IP to every Pod also a concern? You have mentioned about limited IP space but i just wanted to confirm.

**Comment by shahga** [comment]
@aanandr - Yes. Needing IP from underlying VNET/Subnet for every Pod is the worst case - which is what AKS for Windows Containers provides today.

**Comment by johnnyhuy** [comment]
Our main concern with using Azure CNI is that we'd **only** stick to it because it would not be possible to build a cluster instance comprising of different OS nodepools with Kubenet. This would definitely simplify our migration path

**Comment by wangyira** [comment]
Looping in @justindavies and @keikhara here

**Comment by F-Potter** [comment]
Is there any progress yet on allowing windows container for kubenet CNI? Think a lot of people that have IP limitations would love to have this option.

**Comment by F-Potter** [comment]
@wangyira @justindavies @keikhara could you please look at this, this is a major blocker for a large AKS cluster we want to setup. As seen above more people are having this issue.
